### PR TITLE
MM-17759 Add code highlighting

### LINKED ISSUE #3007: Syntax highlight not working
#### Summary

Code-blocks with defined syntax aren't highlighted on iOS app.

#### Environment Information

- Device Name: iPhone (issues exists on Android as well, checked on Samsung galaxy s8)
- OS Version: 12.1
- Mattermost App Version: 1.21
- Mattermost Server Version: 5.12

#### Steps to reproduce

Send message containing syntax highlight, e.g.

````
```php
$a = 'a';
```
````

#### Expected behavior

The syntax highlighted the same way as in webapp.

#### Observed behavior

The block displayed without any highlight.

### PR DESCRIPTION
**Pull request by cardoso**
#### Summary
Implemented code highlighting in code blocks and full-screen code.
Works dynamically with what's defined in the theme configuration (default or custom).

Adds dependency: [react-native-syntax-highlighter](https://github.com/conorhastings/react-native-syntax-highlighter) (React Native wrapper for [react-syntax-highlighter](https://github.com/conorhastings/react-syntax-highlighter))

PS: On Android full-screen code, color is sometimes randomly missing in some parts, but I suspect it's related to how slow the Android emulator runs here.

#### Ticket Link
- GitHub Issue: Closes #3007 
- https://mattermost.atlassian.net/browse/MM-17759

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Has UI changes

#### Device Information
This PR was tested on:
- iPhone SE iOS 12.2
- iPhone 6 -- iOS 12.2 Simulator
- Google Nexus 5X Genymotion Simulator API Version 24

#### Screenshots
| Code Block Light Theme (iOS) | Code Block Light Theme (Android) |
|-|-|
|Screen Shot 2019-08-04 at 12 16 21 PM|Screen Shot 2019-08-04 at 12 18 11 PM|
| Full-screen Code Dark Theme (iOS) | Full-screen Code Dark Theme (Android)* |
|Screen Shot 2019-08-04 at 2 38 38 PM|Screen Shot 2019-08-04 at 2 38 33 PM|
| Code Block Dark Theme (iOS) | Code Block Dark Theme (Android) |
|Screen Shot 2019-08-04 at 1 24 13 PM|Screen Shot 2019-08-04 at 1 24 18 PM|

### CODE REVIEW

File: app/utils/theme.js [hunk 1]
```diff
@@ -42,3 +42,15 @@ export function isThemeSwitchingEnabled(state) {
     const {config} = state.entities.general;
     return config.EnableThemeSelection === 'true';
 }
+
+// getHighlightStyleFromTheme
+
+import * as HighlightStyles from 'react-syntax-highlighter/styles/hljs';
```
Comment (hmhealey) [inline app/utils/theme.js]: Just some style nitpicking, but can you move this up to be with the rest of the import statements?

**Comment by hmhealey** [comment]
Looks excellent! I'm surprised we don't have a Jira ticket for this, but I'll go make one.

I did some testing myself, and outside of the weird highlighting on Android, everything looks good to me. I'll have to test on an actual device to see if it's just a bug in the emulator.

@esethna We might want to consider just shipping this for iOS if it's a concern that the highlighting doesn't always work properly on Android. Here's a comparison of iOS and Android where it's happening:
Screen Shot 2019-08-13 at 9 06 02 AM
Screen Shot 2019-08-13 at 9 05 59 AM

File: app/utils/theme.js (hunk 1 above)
Comment (cardoso) [reply app/utils/theme.js]: Done 👍

**Comment by hmhealey** [comment]
Ticket filed: https://mattermost.atlassian.net/browse/MM-17759

File: app/utils/theme.js [hunk 2]
```diff
@@ -42,3 +43,13 @@ export function isThemeSwitchingEnabled(state) {
     const {config} = state.entities.general;
     return config.EnableThemeSelection === 'true';
 }
+
+// getHighlightStyleFromTheme
```
Comment (saturninoabril) [inline app/utils/theme.js]: Please remove

File: package.json [hunk 3]
```diff
@@ -55,6 +55,7 @@
     "react-native-slider": "0.11.0",
     "react-native-status-bar-size": "0.3.3",
     "react-native-svg": "9.4.0",
+    "react-native-syntax-highlighter": "2.0.1",
```
Comment (saturninoabril) [inline package.json]: Please add entry to https://github.com/mattermost/mattermost-mobile/blob/master/NOTICE.txt

Comment (cardoso) [reply package.json]: Done 👍

File: app/utils/theme.js (hunk 2 above)
Comment (cardoso) [reply app/utils/theme.js]: Done :+1:

**Comment by saturninoabril** [comment]
@esethna Please take a look

File: app/screens/code/code.js [hunk 4]
```diff
@@ -70,21 +75,28 @@ export default class Code extends React.PureComponent {
...
         } else {
             textComponent = (
-                <Text
-                    selectable={true}
-                    style={style.codeText}
+                <SyntaxHighlighter
+                    language={this.props.language}
+                    style={getHighlightStyleFromTheme(this.props.theme)}
+                    highlighter={'hljs'}
+                    customStyle={{...style.codeText, ...this.props.textStyle}}
```
Comment (hmhealey) [inline app/screens/code/code.js]: This line is constructing a new object that gets passed in as the `customStyle` prop each time. Perhaps this is causing the `SyntaxHighlighter` to re-render too often, causing the lag? If we don't create a new object each time, that could possibly help here.

The same thing applies to the line below for `codeTagProps`.

Comment (cardoso) [reply app/screens/code/code.js]: I'll investigate. Thanks!

**Comment by cardoso** [comment]
@hmhealey I found two approaches the highlighter can take:

- split lines into multiple elements (default) - current approach for Android as of last commit
    - doesn't allow text selection
    - color is fine on Android (maybe performance as well?)
- single text element - current approach for iOS
    - allows text selection
    - causes color and performance issues on Android

If the performance issue is fixed on Android with the first approach, we're left with the decision of merging and opening an issue for text selection on Android full-screen code, or simply withholding highlighting on Android full-screen code until a later version of RN fixes this difference in behavior/performance.

**Comment by hmhealey** [comment]
Sorry for the late update on this, @cardoso. I've had a busy week, so I haven't had a chance to test this out yet. I'll take a look at this on Monday to see if this fixes the performance issues on Android or if we're going to have to make this an iOS-only feature for now

**Comment by hmhealey** [comment]
I talked with Eric, and we'd rather not lose the ability to select text on Android. We'd like to support both that and syntax highlighting on Android, but if that doesn't work, we'd be fine with just supporting it on iOS for now.

If you want to take another look at getting it working on Android without the performance issues, feel free to do that, but since it's hard for you to confirm that yourself, feel free to just revert to the old behaviour on Android, and we can create a separate ticket to revisit this on Android

**Comment by enahum** [comment]
@hmhealey can you review this again?

**Comment by cardoso** [comment]
I tried a bunch of things, even forked the library to tweak around, but had to revert the behavior on Android full-screen code as agreed upon.

Highlighting still 100% on iOS.

**Comment by esethna** [comment]
Thanks @cardoso, I'm okay taking this just on iOS if we remove it and create a follow-up ticket for implementing this on Android.

Are there any follow-ups required on your end to remove it from Andoird?

**Comment by cardoso** [comment]
@esethna it's already removed on Android on the screen that causes problems (full-screen code). It's not removed in the message body where it's not causing issues.

Should I remove the highlighting completely from Android?

**Comment by esethna** [comment]
Okay thanks, sounds good as is. We'll get QA to give it a pass

**Comment by esethna** [comment]
@enahum can we build this for testing?

**Comment by enahum** [comment]
@esethna this PR needs changes and update the package-lock.json file before it can be built.

**Comment by cardoso** [comment]
Done 👍

**Comment by enahum** [comment]
@cardoso package-lock.json is adding a bunch of babel 6 references... can you check this? also builds are failing

**Comment by cardoso** [comment]
@enahum I think I fixed the issues, but I haven't tried running the project yet because I won't have time until Saturday.

File: app/components/markdown/markdown_code_block/markdown_code_block.js [hunk 5]
```diff
@@ -165,9 +169,14 @@ export default class MarkdownCodeBlock extends React.PureComponent {
                     </View>
                     <View style={style.rightColumn}>
                         <View style={style.code}>
-                            <Text style={[style.codeText, this.props.textStyle]}>
+                            <SyntaxHighlighter
```
Comment (hmhealey) [inline app/components/markdown/markdown_code_block/markdown_code_block.js]: Can you remove the syntax highlighting from this view on Android? While we can technically support it here, it'd better to be consistent with how we handle it across the the app for them.

Comment (cardoso) [reply app/components/markdown/markdown_code_block/markdown_code_block.js]: Sure. Will do it this weekend.

**Comment by hmhealey** [comment]
@saturninoabril Are you able to re-test this since it's had some decently large changes? Syntax highlighting should be fully supported on iOS, but not on Android.

**Comment by saturninoabril** [comment]
Sorry for the delay.  I'll test within today.

**Pull request merged**